### PR TITLE
Fix bugs and test

### PR DESCRIPTION
**Pull request by gpbp**
I fixed the bug in the method parsePropertiesMetadata and parseProperties in the class GeoJsonReaderDriver by adding four new methods, parseObjectMetadata, parseObject, parseArrayMetadata and parseArray. "parseObject" will ignore hierarchical objects and return "{}".

### CODE REVIEW

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 1]
```diff
@@ -568,5 +568,19 @@ public void testReadComplexFile() throws Exception {
         stat.execute("DROP TABLE IF EXISTS TABLE_POINTS_READ");
         stat.close();
     }
+
+    @Test
+    public void testReadProperties() throws Exception {
+        Statement stat = connection.createStatement();
+        stat.execute("DROP TABLE IF EXISTS TABLE_PROPERTIES_READ;");
+        stat.execute("CALL GeoJsonRead('D:\\users\\pham\\Desktop\\data.geojson','TABLE_PROPERTIES_READ');");
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: The test won't work if the geojson file is not included in the resource folder.

**Comment by SPalominos** [comment]
The travis check should have failed. I will take a look to find why.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 1 above)
Comment (gpbp) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: sorry, i didn't know how to put files into the resource folder.

Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: The resource folder should be place in the maven `resource` folder. Then the file can be accessed by using `ClassName.class.getResource(String relativeResourcePath)` which return the URL of the resource. Be careful, the base of the relative resource path is the path of the class used.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 2]
```diff
@@ -568,5 +568,19 @@ public void testReadComplexFile() throws Exception {
         stat.execute("DROP TABLE IF EXISTS TABLE_POINTS_READ");
         stat.close();
     }
+
+    @Test
+    public void testReadProperties() throws Exception {
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: You should cover all your modifications with test, for example, the method `parseObjectMetadata(JsonParser jp)` of the class GeojsonReaderDriver is not covered.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 3]
```diff
@@ -568,5 +568,19 @@ public void testReadComplexFile() throws Exception {
...
+        Statement stat = connection.createStatement();
+        stat.execute("DROP TABLE IF EXISTS TABLE_PROPERTIES_READ;");
+        stat.execute("CALL GeoJsonRead("+ StringUtils.quoteStringSQL(GeojsonImportExportTest.class.getResource("data.geojson").getPath()) + ", 'TABLE_PROPERTIES_READ');");
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: It is not useful to use a such big file (`data.geojson`) if you test only the very first line. Maybe a smaller file will be better.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 4]
```diff
@@ -568,5 +568,19 @@ public void testReadComplexFile() throws Exception {
...
+        ResultSet res = stat.executeQuery("SELECT * FROM TABLE_PROPERTIES_READ;");
+        res.next();
+        assertEquals(-105576, res.getDouble(2), 0);
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: You should test all the object from the ResultSet to ensure that the geojson have been well parsed.
In this test the property field `mean_pleasantness` is not parsed and the fields `first_measure_ISO_8601` and `last_measure_ISO_8601` should be parsed as timestamp and not string.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java [hunk 5]
```diff
@@ -1329,14 +1338,18 @@ private void parseArrayMetadata(JsonParser jp) throws IOException {
...
      */
-    private void parseObjectMetadata(JsonParser jp) throws IOException {
-        JsonToken value;
-        while(jp.nextToken() != JsonToken.END_OBJECT) {
-            value = jp.nextToken();
-            if (value == JsonToken.START_OBJECT) {
-                parseObjectMetadata(jp);
-            } else if (value == JsonToken.START_ARRAY) {
-                parseArrayMetadata(jp);
+    private void parseObjectMetadata(JsonParser jp, boolean notNull) throws IOException {
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java]: Why did you add the notNull boolean ? In you function, if notNull is true, it doesn't do anything, so why not test in null or not and if not null, call the function ?

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java [hunk 6]
```diff
@@ -1329,14 +1338,18 @@ private void parseArrayMetadata(JsonParser jp) throws IOException {
      * @param jp the json parser
      * @return the object but written like a String
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java]: If you change the method signature, please change its header.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java [hunk 7]
```diff
@@ -461,7 +461,14 @@ private void writeProperties(JsonGenerator jsonGenerator, ResultSet rs) throws I
             for (Map.Entry<String, Integer> entry : cachedColumnNames.entrySet()) {
                 String string = entry.getKey();
                 Integer fieldId = entry.getValue();
-                jsonGenerator.writeObjectField(string, rs.getObject(fieldId));
+                // System.out.println(rs.getObject(fieldId).getClass().getSimpleName());
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: Do not keep commented line if it is not necessary

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java [hunk 8]
```diff
@@ -461,7 +461,14 @@ private void writeProperties(JsonGenerator jsonGenerator, ResultSet rs) throws I
...
+                if (rs.getObject(fieldId) instanceof Array) {
+                    jsonGenerator.writeObjectFieldStart(string);
+                    writeArray(jsonGenerator);
+                    System.out.println("hello");
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: Do not keep System.out.println() if it is not necessary.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 3 above)
Comment (ebocher) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: +1

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java (hunk 6 above)
Comment (ebocher) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java]: +1

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 3 above)
Comment (gpbp) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: @ebocher what do you mean??

**Comment by ebocher** [comment]
By the way, when you commit keep in mind the last comments  
e.g Never keep System.out.println() in the code
e.g Update the documentation
eg Add all necessary tests
eg Do not used large file in test

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 4 above)
Comment (nicolas-f) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: Date cannot be parsed in Json because Date is not a datatype in this format. There is no technical way to identify a date field (trying to check all properties with date regex is not recommended)

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java (hunk 8 above)
Comment (nicolas-f) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: Yes please always debug using line breakpoint in your IDE.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 4 above)
Comment (gpbp) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: so will we leave those fields type be "VARCHAR"?

Comment (gpbp) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: By the way, mean_pleasantness is ignored 'cause its value is null. You guys have idea how to parse null type?

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java [hunk 9]
```diff
@@ -800,7 +808,20 @@ private void parseProperties(JsonParser jp, Object[] values) throws IOException,
                 values[cachedColumnIndex.get(fieldName)] =  jp.getValueAsDouble();
             } else if (value == JsonToken.VALUE_NUMBER_INT) {
                 values[cachedColumnIndex.get(fieldName)] =  jp.getBigIntegerValue();
-            } else {
+            } else if (value == JsonToken.START_ARRAY) {
+                ArrayList<Object> arrayList = parseArray(jp);
+                Object[] array = new Object[arrayList.size()];
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java]: You can do simpler by using `List.toArray()` method which return the Object array corresponding to the list.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java [hunk 10]
```diff
@@ -1278,4 +1297,100 @@ private void setGeometryTypeConstraints() throws SQLException {
...
+        ArrayList<Object> ret = new ArrayList<>();
+        while(value != JsonToken.END_ARRAY) {
+            if (value == JsonToken.START_OBJECT) {
+                Object object = parseObject(jp);
+                ret.add(object);
+            } else if (value == JsonToken.START_ARRAY) {
+                ArrayList<Object> array = parseArray(jp);
+                ret.add(array);
+            } else {
+                ret.add(jp.getValueAsString());
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonReaderDriver.java]: What if the `jp` value is not a `String` bit a `int` for example? Maybe you should take the type into account.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java [hunk 11]
```diff
@@ -50,7 +50,7 @@
  * "geometry":{"type": "Point", "coordinates": [102.0, 0.5]}, "properties":
  * {"prop0": "value0"} } ]}
  *
- * @author Erwan Bocher
+ * @author Erwan Bocher, Hai Trung Pham
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: Same thing for the authors.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java [hunk 12]
```diff
@@ -487,6 +492,8 @@ public boolean isSupportedPropertyType(int sqlTypeId, String sqlTypeName) throws
             case Types.VARCHAR:
             case Types.NCHAR:
             case Types.CHAR:
+            case Types.ARRAY:
+            case Types.TIMESTAMP:
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: It seems that the TIMESTAMP is not supported yet.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java [hunk 13]
```diff
@@ -512,5 +519,17 @@ private void writeCRS(JsonGenerator jsonGenerator, String[] authorityAndSRID) th
...
-    
+
+    /**
+     * Write the array in the geojson
+     *
+     * @param jsonGenerator
+     * @throw IOException
+     * @author Pham Hai Trung
+     */
+    private void writeArray(JsonGenerator jsonGenerator) throws IOException, SQLException {
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: Where do you add the array content ?

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 14]
```diff
@@ -37,7 +37,7 @@
 
 /**
  *
- * @author Erwan Bocher
+ * @author Erwan Bocher, Hai Trung Pham
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: Same thing for the authors.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 4 above)
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: For now, let the fields in `VARCHAR`.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java [hunk 15]
```diff
@@ -512,5 +519,16 @@ private void writeCRS(JsonGenerator jsonGenerator, String[] authorityAndSRID) th
...
+
+    /**
+     * Write the array in the geojson
+     *
+     * @param jsonGenerator
+     * @throw IOException
+     */
+    private void writeArray(JsonGenerator jsonGenerator) throws IOException, SQLException {
+        jsonGenerator.writeStartArray();
+        jsonGenerator.writeEndArray();
```
Comment (SPalominos) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: This method only write  the start and the end of the array, but no its content.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 16]
```diff
@@ -568,5 +570,31 @@ public void testReadComplexFile() throws Exception {
...
+        assertEquals(55.87291487481895, res.getDouble(5), 0);
+        assertEquals(0.0, res.getDouble(6), 0);
+        assertEquals(2, res.getDouble(7), 0);
+        assertTrue(res.getString(8).equals("2017-01-19T18:29:26+01:00"));
+        assertTrue(res.getBigDecimal(9).toString().equals("1484846966000"));
+        assertTrue(res.getString(10).equals("2017-01-19T18:29:27+01:00"));
+        assertTrue(res.getBigDecimal(11).toString().equals("1484846967000"));
+        assertTrue(res.getString(12).equals("{}"));
+        //assertTrue((res.getArray(13)).equals());
+        //assertTrue((res.getArray(14)).equals(WKTREADER.read("(58, 47, 58, 57, 58, 49, 58, 51, 58, 58, 49, 57, 58, 58, 49, 58, 57, 56, 57, 58, 59, 58, 57, 58, 49, 47, 48, 57, 48, 58, 57, 57, 51, 56, 52, 57, 51, 57, 49, 58, 55, 58, 50, 48, 48, 52, 56, 57, 48, 58, 52, 48, 53, 50, 57, 54, 57, 47, 58, 57, 54, 54, 53, 56, 57, 55, 58, 58, 57, 58, 57, 57)")));
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: Do not forget to implement those two asserts to test the array parsing.

File: h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java (hunk 15 above)
Comment (gpbp) [inline h2gis-functions/src/main/java/org/h2gis/functions/io/geojson/GeoJsonWriteDriver.java]: yeah, I'll create a whole new branch to fix it

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 16 above)
Comment (gpbp) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: Actually I don't know how to test an array so I leave it to ask you guys.

Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: Did you just check the methods provided by the `org.junit.Assert` class ? It contains methods (like `assertArrayEquals`) to test arrays. For more info, check the [junit javadoc](http://junit.org/junit4/javadoc/latest/org/junit/Assert.html).

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java (hunk 3 above)
Comment (ebocher) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: do not push a big file.

**Comment by SPalominos** [comment]
You just have to change the author in the class `GeojsonImportExportTest` and I will be able to approve your changes.

File: h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java [hunk 17]
```diff
@@ -568,5 +570,51 @@ public void testReadComplexFile() throws Exception {
...
+        assertEquals(2, res.getDouble(7), 0);
+        assertTrue(res.getString(8).equals("2017-01-19T18:29:26+01:00"));
+        assertTrue(res.getBigDecimal(9).toString().equals("1484846966000"));
+        assertTrue(res.getString(10).equals("2017-01-19T18:29:27+01:00"));
+        assertTrue(res.getBigDecimal(11).toString().equals("1484846967000"));
+        assertTrue(res.getString(12).equals("{}"));
+        Object[] tinyArray = {(Integer) 13, "string", "{}"};
+        Object[] expectedResult = {(Integer) 49, (Double) 40.0, "{}", "string", tinyArray};
+        Object[] result = (Object[]) res.getObject(13);
+        assertEquals(expectedResult.length, result.length);
```
Comment (SPalominos) [inline h2gis-functions/src/test/java/org/h2gis/functions/io/geojson/GeojsonImportExportTest.java]: Use `assertArrayEquals()` to test arrays.
